data-protocols/ids/ids-api-multipart-dispatcher-v1/src/main/java/org/eclipse/dataspaceconnector/ids/api/multipart/dispatcher/IdsMultipartDispatcherServiceExtension.java

```diff
@@ -80,14 +83,16 @@ public void initialize(ServiceExtensionContext context) {
         var objectMapper = context.getTypeManager().getMapper("ids");
         var typeManager = context.getTypeManager();
 
+        var senderContext = new DelegateMessageContext(URI.create(connectorId), objectMapper, transformerRegistry, idsWebhookAddress);
+
         var dispatcher = new IdsMultipartRemoteMessageDispatcher();
-        dispatcher.register(new MultipartArtifactRequestSender(connectorId, httpClient, objectMapper, monitor, vault, identityService, transformerRegistry, idsWebhookAddress));
-        dispatcher.register(new MultipartDescriptionRequestSender(connectorId, httpClient, objectMapper, monitor, identityService, transformerRegistry));
-        dispatcher.register(new MultipartContractOfferSender(connectorId, httpClient, objectMapper, monitor, identityService, transformerRegistry, idsWebhookAddress));
-        dispatcher.register(new MultipartContractAgreementSender(connectorId, httpClient, objectMapper, monitor, identityService, transformerRegistry, idsWebhookAddress));
-        dispatcher.register(new MultipartContractRejectionSender(connectorId, httpClient, objectMapper, monitor, identityService, transformerRegistry));
-        dispatcher.register(new MultipartCatalogDescriptionRequestSender(connectorId, httpClient, objectMapper, monitor, identityService, transformerRegistry));
-        dispatcher.register(new MultipartEndpointDataReferenceRequestSender(connectorId, httpClient, objectMapper, monitor, identityService, transformerRegistry, typeManager));
+        dispatcher.register(new IdsMultipartSender<>(monitor, httpClient, identityService, objectMapper, new MultipartArtifactRequestSender(senderContext, vault)));
+        dispatcher.register(new IdsMultipartSender<>(monitor, httpClient, identityService, objectMapper, new MultipartDescriptionRequestSender(senderContext)));
+        dispatcher.register(new IdsMultipartSender<>(monitor, httpClient, identityService, objectMapper, new MultipartContractOfferSender(senderContext)));
+        dispatcher.register(new IdsMultipartSender<>(monitor, httpClient, identityService, objectMapper, new MultipartContractAgreementSender(senderContext)));
+        dispatcher.register(new IdsMultipartSender<>(monitor, httpClient, identityService, objectMapper, new MultipartContractRejectionSender(senderContext)));
+        dispatcher.register(new IdsMultipartSender<>(monitor, httpClient, identityService, objectMapper, new MultipartCatalogDescriptionRequestSender(senderContext)));
+        dispatcher.register(new IdsMultipartSender<>(monitor, httpClient, identityService, objectMapper, new MultipartEndpointDataReferenceRequestSender(senderContext, typeManager)));
```

Comment:
In the future would be great to clearly separate the "sender" from the "delegates" so that the dispatcher would be able to choose the sender by the protocol and the delegate by message type.
Not needed now, just to see if we're on the same track.

Reply:
Do you mean that we'd have one dispatcher for IDS messages that keeps a list of senders for the protocol and a list of delegates for the message type? I'm not sure the same delegates can be used for different IDS protocols, as e.g. for IDS REST the IDS message header information would be mapped to HTTP headers. But maybe we could rewrite the `IdsMultipartRemoteMessageDispatcher` to choose the delegate depending on the message type and hand that over to the `IdsMultipartSender` as a parameter, what do you think?

Reply:
@ronjaquensel Maybe I went too far with the speculation, it's something that could raise when another protocol will pop up (if it ever would).
In the meantime I think your proposal to choose the delegate depending on the message type and sending it to the `IdsMultipartSender` makes sense, doing that we we'll end up with a single sender instance and the dispatcher would just choose the delegate.

Reply:
refactored it